librz/core/linux_heap_glibc.c

```diff
@@ -1406,632 +1836,473 @@ static void GH(print_heap_segment)(RzCore *core, MallocState *main_arena,
-					free(dup);
-					break;
-				}
-				GH(print_heap_fastbin)
-				(core, m_state, main_arena, global_max_fast, dup, main_arena_only);
-			} else {
-				PRINT_RA("This address is not part of the arenas\n");
-				free(dup);
-				break;
-			}
-			free(dup);
+RZ_IPI int GH(rz_cmd_heap_fastbins_print)(void *data, const char *input) {
+	RzCore *core = (RzCore *)data;
+	GHT m_arena = GHT_MAX, m_state = GHT_MAX;
+	RzConsPrintablePalette *pal = &rz_cons_singleton()->context->pal;
+	MallocState *main_arena = RZ_NEW0(MallocState);
+	GHT global_max_fast = (64 * SZ / 4);
+	if (!GH(rz_heap_resolve_main_arena)(core, &m_arena)) {
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
+	}
+	bool main_arena_only = false;
+	char *m_state_str, *dup = strdup(input);
+	if (*dup) {
+		strtok(dup, ":");
+		m_state_str = strtok(NULL, ":");
+		m_state = rz_num_get(NULL, m_state_str);
+		if (!m_state) {
+			m_state = m_arena;
 		}
-		break;
-	case 'v':
-		if (input[0] == 'v') {
-			format = 'v';
-		}
-	case 'g': //dmhg
-		if (input[0] == 'g') {
-			format = 'g';
-		}
-	case '*': //dmh*
-		if (input[0] == '*') {
-			format = '*';
-		}
-	case 'j': // "dmhj"
-		if (input[0] == 'j') {
-			format = 'j';
-		}
-		if (GH(rz_resolve_main_arena)(core, &m_arena)) {
-			input += 1;
-			if (!strcmp(input, "\0")) {
-				if (core->offset != core->prompt_offset) {
-					m_state = core->offset;
-					get_state = true;
-				}
-			} else {
-				m_state = rz_num_get(NULL, input);
-				get_state = true;
-			}
-			if (!get_state) {
-				m_state = m_arena;
-			}
-			if (GH(is_arena)(core, m_arena, m_state)) {
-				if (!GH(update_main_arena)(core, m_state, main_arena)) {
-					break;
-				}
-				GH(print_heap_segment)
-				(core, main_arena, m_arena, m_state, global_max_fast, format);
-			} else {
-				PRINT_RA("This address is not part of the arenas\n");
-			}
+	} else {
+		if (core->offset != core->prompt_offset) {
+			m_state = core->offset;
+		} else {
+			m_state = m_arena;
 		}
-		break;
-	case 't':
-		if (GH(rz_resolve_main_arena)(core, &m_arena)) {
-			if (!GH(update_main_arena)(core, m_arena, main_arena)) {
-				break;
-			}
-			bool main_thread_only = false;
-			GH(print_tcache_instance)
-			(core, m_arena, main_arena, main_thread_only);
+	}
+	if (GH(is_arena)(core, m_arena, m_state)) {
+		if (!GH(rz_heap_update_main_arena)(core, m_state, main_arena)) {
+			free(dup);
+			free(main_arena);
+			return RZ_CMD_STATUS_ERROR;
 		}
-		break;
-	case '?':
-		rz_core_cmd_help(core, GH(help_msg));
-		break;
+		GH(print_heap_fastbin)
+		(core, m_state, main_arena, global_max_fast, dup, main_arena_only, NULL);
+	} else {
+		PRINT_RA("This address is not part of the arenas\n");
+		free(dup);
+		free(main_arena);
+		return RZ_CMD_STATUS_ERROR;
 	}
+	free(dup);
 	free(main_arena);
-	return true;
+	return RZ_CMD_STATUS_OK;
 }
+
+RZ_IPI RzCmdStatus GH(rz_cmd_heap_arena_bins_print_handler)(RzCore *core, int argc, const char **argv, RzOutputMode mode) {
```

Comment:
```
[0x55a48dde7d80]> dmhd fastbin 
Tcache in Main Arena @  0x7f2f1eb42a00


Fast bins in Arena @ 0x7f2f1eb42a00
Fast_bin[01] [size: 0x20] Empty bin
Fast_bin[02] [size: 0x30] Empty bin
Fast_bin[03] [size: 0x40] Empty bin
Fast_bin[04] [size: 0x50] Empty bin
Fast_bin[05] [size: 0x60] Empty bin
Fast_bin[06] [size: 0x70] Empty bin
Fast_bin[07] [size: 0x80] Empty bin

Unsorted bin in Arena @ 0x7f2f1eb42a00
Found 0 chunks in unsorted bin

Small bins in Arena @ 0x7f2f1eb42a00
Found 0 chunks in 0 small bins

Large bins in Arena @ 0x7f2f1eb42a00
Found 0 chunks in 0 large bins
```
I would expect only fastbins. Works with other bin_type, but not with fastbins.

Reply:
Currently `dmhd fast` will show fastbins instead of `dmhd fastbin`, I can add both if you want

Reply:
I think you used `fastbin` in the CHOICES of the argument. Either accept `fastbin` correctly in this command or change the CHOICES to use `fast` instead of `fastbin`.

Reply:
Yeah my bad, it should have been "fast", anyways I will change it so that it accepts both

Reply:
I think it's better to change CHOICES to `fast` and do not support `fastbin` to be consistent.